urla42f49a9b2_dpjayasekara_crypthouse_tgz-pJ8-vehiclesc_VehicleSCJ8/src/groupsignature/elliptic/ECPoint.java

```diff
@@ -0,0 +1,210 @@
+	}
+	iszero = false;
+    }
+     
+    /** IMPORTANT this renders the values of x and y to be null! Use this constructor
+     *only to create instances of a Zero class!*/
+    public ECPoint(EllipticCurve e){
+	x = y = BigInteger.ZERO;
+	mother = e;
+	iszero = true;
+    }
+
+    public byte[] compress() {
+	byte[] cmp = new byte[mother.getPCS()];
+	if (iszero){
+	    cmp[0] = 2;
+	}
+	byte[] xb = x.toByteArray();
+	System.arraycopy(xb, 0, cmp, mother.getPCS()-xb.length, xb.length);
+	if(y.testBit(0)) cmp[0] = 1;
+	return cmp;
+    }
+
+    /** Adds another elliptic curve point to this point.
+     *@param q The point to be added
+     *@return the sum of this point on the argument
+     *@exception Throws a NoCommonMotherException if the two points don't lie on the same elliptic curve.*/
+    public ECPoint add(ECPoint q) throws NoCommonMotherException{
+
+	if (!hasCommonMother(q)) throw new NoCommonMotherException();
+
+	if (this.iszero) return q;
+	else if (q.isZero()) return this;
+
+	BigInteger y1 = y;
+	BigInteger y2 = q.gety();
+	BigInteger x1 = x;
+	BigInteger x2 = q.getx();
+
+	BigInteger alpha;
+
+	if (x2.compareTo(x1) == 0) {
+
+	    if (!(y2.compareTo(y1) == 0)) return new ECPoint(mother);
+	    else {
+		alpha = ((x1.modPow(TWO,mother.getp())).multiply(THREE)).add(mother.geta());
+		alpha = (alpha.multiply((TWO.multiply(y1)).modInverse(mother.getp()))).mod(mother.getp());
+	    }
+
+	} else {
+	    alpha = ((y2.subtract(y1)).multiply((x2.subtract(x1)).modInverse(mother.getp()))).mod(mother.getp());
+	} 
+
+	BigInteger x3,y3;
+	x3 = (((alpha.modPow(TWO,mother.getp())).subtract(x2)).subtract(x1)).mod(mother.getp());
+	y3 = ((alpha.multiply(x1.subtract(x3))).subtract(y1)).mod(mother.getp());
+ 
+
+	try{ return new ECPoint(mother,x3,y3); }
+	catch (NotOnMotherException e){
+	    System.out.println("Error in add!!! Result not on mother!");
+	    return null;
+	}
+
+    }
+    
+    public ECPoint multiply(BigInteger coef) {
+	try{
+	    ECPoint result = new ECPoint(mother);
+	    byte[] coefb = coef.toByteArray();
+	    if(fastcache != null) {
+		for(int i = 0; i < coefb.length; i++) {
+		    result = result.times256().add(fastcache[coefb[i]&255]);
+		}
+		return result;
+	    }
+	    if(cache == null) {
+		cache = new ECPoint[16];
+		cache[0] = new ECPoint(mother);
+		for(int i = 1; i < cache.length; i++) {
+		    cache[i] = cache[i-1].add(this);
+		}
+	    }
+	    for(int i = 0; i < coefb.length; i++) {
+		result = result.times16().add(cache[(coefb[i]>>4)&15]).times16().add(cache[coefb[i]&15]);
+	    }
+	    return result;
+	} catch (NoCommonMotherException e) {
+	    System.out.println("Error in pow!!!");
+	    return null;
+	}
+    }
+
+    private ECPoint times16() {
+	try {
+	    ECPoint result = this;
+	    for(int i = 0; i < 4; i++) {
+		result = result.add(result);
+	    }
+	    return result;
+	} catch (Exception e) {
+	    System.out.println("ECPoint.times16: THIS CANNOT HAPPEN!!!");
+	    return null;
+	}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 161:170, 174:183.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.